src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesVsPatterns.cs

```diff
@@ -1933,5 +1933,81 @@ public interface IOut<out T> { }
                 //         _ = i switch { 1 => default, _ => x }/*T:T*/; // 6
                 Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "default").WithLocation(46, 29));
         }
+
+        [Fact]
+        [WorkItem(39264, "https://github.com/dotnet/roslyn/issues/39264")]
+        public void IsPatternSplitState_01()
+        {
+            CSharpCompilation c = CreateNullableCompilation(@"
+#nullable enable
+class C
+{
+    string? field = string.Empty;
+
+    void M1(C c)
+    {
+        if (c.field == null) return;
+        
+        c.field.ToString();
+    }
+    
+    void M2(C c)
+    {
+        if (c is { field: null }) return;
+        
+        c.field.ToString();
+    }
+
+    void M3(C c)
+    {
+        switch (c)
+        {
+            case { field: null }:
+                break;
+            default:
+                c.field.ToString();
+                break;
+        }
+    }
+
+    void M4(C c)
+    {
+        _ = c switch
+        {
+            { field: null } => string.Empty,
+            _ => c.field.ToString(),
+        };
+    }
+}
+");
+            c.VerifyDiagnostics(
+                );
+        }
+
+        [Fact]
+        [WorkItem(39264, "https://github.com/dotnet/roslyn/issues/39264")]
+        public void IsPatternSplitState_02()
+        {
+            CSharpCompilation c = CreateNullableCompilation(@"
+#nullable enable
+class C
+{
+    C? c = null;
+
+    void M1(C c, bool b)
+    {
+        if (c is { c : null })
+        {
+            if (c.c != null)
+            {
+                c.c.c.c.ToString();
```

Comment:
Why is there no warning here? We tested `c.c`, but not `c.c.c` and not `c.c.c.c`.

Reply:
That is the point of this change.  We don't need a warning, because the compiler has inferred that null cannot arise here as a consequence of previous tests.  (Try to write a counterexample.  You cannot.)